modules/fluid_properties/include/utils/NewtonInversion.h

```diff
@@ -19,18 +20,73 @@ namespace FluidPropertiesUtils
  * @param[in] x constant first argument of the f(x, z) term
  * @param[in] y constant which should be equal to f(x, z) with a converged z
  * @param[in] z_initial_guess initial guess for return variables
- * @param[in] tolerance criterion on absolute difference between successive iterates to judge
- * convergence
+ * @param[in] tolerance criterion for relative or absolute (if y is sufficently close to zero)
+ * convergence checking
  * @param[in] function two-variable function returning both values and derivatives as references
  * @param[in] max_its the maximum number of iterations for Newton's method
- * @return the value z such that f(x, z) = y
+ * @return a pair in which the first member is the value z such that f(x, z) = y and the second
+ * member is dy/dz
  */
-Real NewtonSolve(const Real & x,
-                 const Real & y,
-                 const Real & z_initial_guess,
-                 const Real & tolerance,
-                 std::function<void(Real, Real, Real &, Real &, Real &)> const & func,
-                 const unsigned int max_its = 100);
+template <typename T, typename Functor>
+std::pair<T, T>
+NewtonSolve(const T & x,
+            const T & y,
+            const Real z_initial_guess,
+            const Real tolerance,
+            const Functor & func,
+            const unsigned int max_its = 100)
+{
+  // R represents residual
+
+  std::function<bool(const T &, const T &)> abs_tol_check =
+      [tolerance](const T & R, const T & /*y*/)
+  { return MetaPhysicL::raw_value(std::abs(R)) < tolerance; };
+  std::function<bool(const T &, const T &)> rel_tol_check = [tolerance](const T & R, const T & y)
+  { return MetaPhysicL::raw_value(std::abs(R / y)) < tolerance; };
+  auto convergence_check = MooseUtils::absoluteFuzzyEqual(MetaPhysicL::raw_value(y), 0, tolerance)
+                               ? abs_tol_check
+                               : rel_tol_check;
+
+  T z = z_initial_guess, R, new_y, dy_dx, dy_dz;
+  unsigned int iteration = 0;
+
+  do
+  {
+    func(x, z, new_y, dy_dx, dy_dz);
+    R = new_y - y;
+
+    // We always want to perform at least one update in order to get derivatives on z correct (z
+    // corresponding to the initial guess will have no derivative information), so we don't
+    // immediately return if we are converged
+    const bool converged = convergence_check(R, y);
+
+#ifndef NDEBUG
+    static constexpr Real perturbation_factor = 1 + 1e-8;
+    T perturbed_y, dummy, dummy2;
+    func(x, perturbation_factor * z, perturbed_y, dummy, dummy2);
+    // Check the accuracy of the Jacobian
+    auto J_differenced = (perturbed_y - new_y) / (1e-8 * z);
+    if (!MooseUtils::relativeFuzzyEqual(J_differenced, dy_dz, 1e-2))
+      mooseError("Bad Jacobian in NewtonSolve");
```

Comment:
```suggestion
      mooseWarning("Bad Jacobian in NewtonSolve");
```

some derivatives are implemented by finite difference, so we could hit this in acceptable conditions